src/script/interpreter.cpp

```diff
@@ -1291,35 +1379,174 @@ uint256 GetOutputsSHA256(const T& txTo)
+        if (!txTo.vin[inpos].scriptWitness.IsNull()) {
+            if (m_spent_outputs_ready && m_spent_outputs[inpos].scriptPubKey.size() == 2 + WITNESS_V1_TAPROOT_SIZE &&
+                m_spent_outputs[inpos].scriptPubKey[0] == OP_1) {
+                // Treat every native witness v1 spend as a Taproot spend. This only works if spent_outputs was
+                // provided as well, but if it wasn't, actual validation will fail anyway.
+                uses_bip341 = true;
+            } else {
+                // Treat every spend that's not known to native witness v1 as a Witness v0 spend. This branch may
+                // also be taken for unknown witness versions, but it is harmless, and being precise would require
+                // P2SH evaluation to find the redeemScript.
+                uses_bip143 = true;
+            }
+        }
+    }
+
+    if (uses_bip143 || uses_bip341) {
+        // Computations shared between both sighash schemes.
+        m_bip341_prevouts_hash = GetPrevoutsSHA256(txTo);
+        m_bip341_sequences_hash = GetSequencesSHA256(txTo);
+        m_bip341_outputs_hash = GetOutputsSHA256(txTo);
+    }
+    if (uses_bip143) {
+        hashPrevouts = SHA256Uint256(m_bip341_prevouts_hash);
+        hashSequence = SHA256Uint256(m_bip341_sequences_hash);
+        hashOutputs = SHA256Uint256(m_bip341_outputs_hash);
+        m_bip143_ready = true;
+    }
+    if (uses_bip341) {
+        m_bip341_spent_amounts_hash = GetSpentAmountsSHA256(m_spent_outputs);
+        m_bip341_spent_scripts_hash = GetSpentScriptsSHA256(m_spent_outputs);
+        m_bip341_ready = true;
+    }
 }
 
 template <class T>
 PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
 {
-    Init(txTo);
+    Init(txTo, {});
 }
 
 // explicit instantiation
-template void PrecomputedTransactionData::Init(const CTransaction& txTo);
-template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo);
+template void PrecomputedTransactionData::Init(const CTransaction& txTo, std::vector<CTxOut> spent_outputs);
+template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo, std::vector<CTxOut> spent_outputs);
 template PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo);
 template PrecomputedTransactionData::PrecomputedTransactionData(const CMutableTransaction& txTo);
 
+static const CHashWriter HASHER_TAPSIGHASH = TaggedHash("TapSighash");
+static const CHashWriter HASHER_TAPLEAF = TaggedHash("TapLeaf");
+static const CHashWriter HASHER_TAPBRANCH = TaggedHash("TapBranch");
+static const CHashWriter HASHER_TAPTWEAK = TaggedHash("TapTweak");
+
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata, const T& tx_to, const uint32_t in_pos, const uint8_t hash_type, const SigVersion sigversion, const PrecomputedTransactionData* cache)
+{
+    uint8_t ext_flag;
+    switch (sigversion) {
+    case SigVersion::TAPROOT:
+        ext_flag = 0;
```

Comment:
I think BIP341 footnote mentions that TAPSCRIPT's `ext_flag=1` but I don't find where it mandates that TAPROOT's one is actually `ext_flag=0`

Reply:
BIP341 says to use `0x00 || SigMsg(0x00, 0)` as message. The `0` there is the `ext_flag` argument.